design/0001-frr.md

```diff
@@ -0,0 +1,347 @@
+                              │               │  │               │
+                              └───────────────┘  └───────┬───────┘
+                                                         │
+                                                         │  <<use>>
+                                                 ┌───────▼───────┐
+                                                 │               │
+                                                 │    session    │
+                                                 │ <<interface>> │
+                                                 │               │
+                                                 └───────▲───────┘
+                                                         │ <<implement>>
+                                             ┌───────────┴─────────────┐
+                                             │                         │
+                                     ┌───────┴───────┐         ┌───────┴───────┐
+                                     │               │         │               │
+                                     │  bgp_metallb  │         │    bgp_frr    │
+                                     │  <<package>>  │         │  <<package>>  │
+                                     │               │         │               │
+                                     └───────────────┘         └───────┬───────┘
+                                                                       │
+                                                                       │
+                                                               ┌───────┴───────┐
+                                                               │               │
+                                                               │      frr      │
+                                                               │ <<container>> │
+                                                               │               │
+                                                               └───────────────┘
+
+```
+
+#### FRR deployment
+
+FRR will be deployed as a container. The speaker application will be configured
+at startup to use the FRR bgp daemon and will provide details for connection to
+the daemon (e.g. location of a socket)
+
+The container will expose a control interface that will initially be used to
+configure FRR via the `vtysh` application and eventually via its gRPC golang
+bindings. This will be used to implement bgp.New() and bgp.Close() for session
+creation and bgp.Set() for prefix advertisement.
+
+#### others
+
+There may be some common functionality in the bgp package that could be reused
+between BGP implementations. For example, the new bgp package will require
+integration with the “metrics” struct which updates prometheus in response to
+BGP events. It may be necessary to refactor these types of functionality into a
+separate package for reuse amongst BGP implementations.
+
```

Comment:
what about exising limitation betweem MetallB BGP and any external component that can also do BGP peering like the following 
- https://metallb.universe.tf/configuration/calico/
- https://metallb.universe.tf/configuration/kube-router/
will moving to FRR help fix the above issues/limitations ?

Reply:
If the current native BGP agent peers with an external router, this will preclude another BGP agent on that node peering with the same router. This is due to BGP’s conflict resolution algorithm as stated in the links above. This is a limitation with the native BGP implementation in MetalLB.

At an architectural level, allowing MetalLB to reuse/share a BGP agent that is already present on the node would appear to resolve this issue as we would let the BGP agent resolve conflicts and act as the single point of contact with the peered router. However, there may be implementation details that would prevent this. Also, the ability to share the BGP agent is currently not an explicit goal of this enhancement. I think that as development of this feature continues, it should be something to investigate and I will update this enhancement accordingly. Thanks @msherif1234!